### DISCUSSION TITLE
Can @Retryable use withinMillis option?

### DISCUSSION THREAD

**Discussion by Alessandrovito** [spring-projects/spring-retry#305] [Q&A]
I'm interesting to have retry based on timeout.
So I created the following fluent API:

> RetryTemplate retryTemplateOk =  RetryTemplate.builder()
>                 .withinMillis(50000)
>                 .fixedBackoff(1000)
>                 .retryOn(Arrays.asList(OperationCancellationException.class, AmbiguousTimeoutException.class))
>                 .build()

In the annotation ` @Retryable` we have only the option `maxAttempts` so looks like we  cannot create a retry based on timeout but only based on attempts. Is this correct ? In case have you already planned to add this support in the annotation ?

**Comment by garyrussell** [comment]
The builder uses a `TimeoutRetryPolicy`; it is not directly supported by `@Retryable`.

You can configure your own `RetryOperationsInterceptor` bean and set it on `@Retryable` in the `interceptor` property.

**Reply by Alessandrovito** [reply]
Good recommandation, but using the `RetryOperationsInterceptor.builder` to create my own `RetryOperationsInterceptor`  bean  it is not possible to configure at the same time RetryPolicy and backOffPolicy. 
So now if I configure only the RetryPolicy into the bean  `RetryOperationsInterceptor` and try to enforce the backOffPolicy as option in the @Retryable` like the following:

> backoff = @Backoff(delay = 1000)

then it doesn't take any effect, it is ignored and the retry it is executed as soon as possible. Any suggestion ?
This is my method interface:

>  @Retryable(value = { OperationCancellationException.class, AmbiguousTimeoutException.class  }, backoff = @Backoff(delay = 1000),  interceptor = "serviceQueryRetryInterceptor", recover = "getOffsetForPartitionsInRepoWithTheSupportOfRetryMechanismResponseFallback" )

and this is my bean `RetryOperationsInterceptor`

>      @Bean(name = "serviceQueryRetryInterceptor")
>     public RetryOperationsInterceptor serviceQueryRetryInterceptor() {
> 
>         TimeoutRetryPolicy retryPolicy = new TimeoutRetryPolicy();
>         retryPolicy.setTimeout(30000);
> 
>         final FixedBackOffPolicy fixedBackOffPolicy = new FixedBackOffPolicy();
>         fixedBackOffPolicy.setBackOffPeriod(1000);
> 
>         return RetryInterceptorBuilder.stateless()
>                 //.backOffPolicy(fixedBackOffPolicy)
>                 .retryPolicy(retryPolicy).build();
>     }

**Reply by artembilan** [reply]
Try to flip those options:
```
.retryPolicy(retryPolicy)
.backOffPolicy(fixedBackOffPolicy)
```

See the logic in their source code:
```

	/**
	 * Apply the retry policy - cannot be used if a custom retry template has been
	 * provided, or the max attempts or back off options or policy have been applied.
	 * @param policy The policy.
	 * @return this.
	 */
	public RetryInterceptorBuilder<T> retryPolicy(RetryPolicy policy) {
		Assert.isNull(this.retryOperations, "cannot set the retry policy when a custom retryOperations has been set");
		Assert.isTrue(!this.templateAltered,
				"cannot set the retry policy if max attempts or back off policy or options changed");
		this.retryTemplate.setRetryPolicy(policy);
		this.retryPolicySet = true;
		this.templateAltered = true;
		return this;
	}

	/**
	 * Apply the back off policy. Cannot be used if a custom retry operations, or back off
	 * policy has been applied.
	 * @param policy The policy.
	 * @return this.
	 */
	public RetryInterceptorBuilder<T> backOffPolicy(BackOffPolicy policy) {
		Assert.isNull(this.retryOperations,
				"cannot set the back off policy when a custom retryOperations has been set");
		Assert.isTrue(!this.backOffOptionsSet,
				"cannot set the back off policy when the back off policy options have been set");
		this.retryTemplate.setBackOffPolicy(policy);
		this.templateAltered = true;
		this.backOffPolicySet = true;
		return this;
	}
```

So, `backOffPolicy` does change `templateAltered`, but it does not check for it.
Not sure, though, why the logic has been chosen like that: doesn't look like a `RetryTemplate` has a problem setting both `BackOffPolicy` and `RetryPolicy`...

**Reply by garyrussell** [reply]
When you use `interceptor` all other properties are ignored - see the Javadoc.

Use `builder.retryOperations(fullyAssembledRetryTemplateWithRetryAndBackOffPolicies)`.

**Reply by Alessandrovito** [reply]
Great now it works! I can confirm the `RetryTemplate` doesn't have this kind of problem . I'm using the fluent API thought the `RetryTemplate.builder()`
